File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.js
// function solution(expression) {
//     // 가능한 우선순위 조합
//     const prior = [
//         ['+', '-', '*'],
//         ['+', '*', '-'],
//         ['-', '+', '*'],
//         ['-', '*', '+'],
//         ['*', '-', '+'],
//         ['*', '+', '-']
//     ]
//     let result = [] // 각 우선순위에 따른 계산한 결과를 담아두는 배열
    
//     for (let opArr of prior) { // opArr = ['+', '-', '*']
//         const temp = expression.split(/(\D)/) 
//         // [ '100', '-', '200', '*', '300', '-', '500', '+', '20']
//         for (let op of opArr) { // op = '-'
//             while (temp.includes(op)) {
//                 const idx = temp.indexOf(op) // op의 앞의 숫자, op, op뒤의 숫자를 자르고, 뒤의 걸로 대체한다.
//                 const cal = calculate(temp[idx - 1], op, temp[idx + 1])
//                 temp.splice(idx - 1, 3, cal)
//             }
//         }
//         result.push(Math.abs(temp[0]))
//     }
//     return Math.max(...result)
// }
    
// function calculate(a, op, b) {
//     if (op === '+') return +a + +b
//     if (op === '-') return a - b
//     if (op === '*') return a * b
// }

function solution(expression) {
    const expressions = [
        ['+', '-', '*'],
        ['+', '*', '-'],
        ['-', '+', '*'],
        ['-', '*', '+'],
        ['*', '-', '+'],
        ['*', '+', '-']
    ]
    let max = 0
        
    let tokens = expression.match(/(\d+|[-+*/])/g);
    // ['100', '-','200', '*', '300', '-', '500', '+', '20']
    
    expressions.forEach((ex, i) => { // '+-*'
        let tokens_c = [...tokens] // ['100', '-','200', '*', '300', '-', '500', '+', '20']
        console.log('ex', ex, 'i', i)
        ex.forEach((e) => { // e = '+'
            for (let i = 0; i < tokens_c.length; i++) {
                if (e === tokens_c[i]) { // tokens_c[i] = '+'
                    let temp;
                    if (e === '+') {
                        temp = +tokens_c[i - 1] + +tokens_c[i + 1];
                    } else if (e === '*') {
                        temp = +tokens_c[i - 1] * +tokens_c[i + 1];
                    } else {
                        temp = +tokens_c[i - 1] - +tokens_c[i + 1];
                    }
                    tokens_c.splice(i - 1, 3, temp.toString());
                    i--;
                } 
            }
        });
        max = Math.max(max, Math.abs(Number(tokens_c[0])));
    });
    
    return max
}

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
import re

def solution(expression): # 연산자(+,-,*)의 우선순위를 재정의하여 절댓값이 가장 큰 숫자 만들기
    answer = []
    cases = [
        ['+', '-', '*'],
        ['+', '*', '-'],
        ['-', '+', '*'],
        ['-', '*', '+'],
        ['*', '+', '-'],
        ['*', '-', '+']
    ]
    
    for case in cases: # 연산자 우선순위 배열
        tokens = re.split(r'(\D)', expression)  # 숫자가 아닌 문자 기준으로 분리
        # ['100', '-', '200', '*', '300', '-', '500', '+', '20']
        for operator in case: # 처리중인 연산자
            while operator in tokens:
                # 연산자가 있는 위치를 찾음
                op_idx = tokens.index(operator)
                
                # 연산자 앞뒤 숫자와 연산자를 추출
                a = tokens[op_idx - 1]
                b = tokens[op_idx + 1]
                
                # 계산
                calculated = eval(f"{a} {operator} {b}")

                 # tokens 리스트 업데이트: 계산된 값 반영
                tokens = tokens[:op_idx - 1] + [str(calculated)] + tokens[op_idx + 2:]      
        
        [acc_calculated] = tokens
        answer.append(abs(int(acc_calculated)))
        
    return max(answer)